### PR TITLE
SOC-605 Implement blog and article comments using the new Email extension

### PR DESCRIPTION
**Pull request by garthwebb**
https://wikia-inc.atlassian.net/browse/SOC-605

### CODE REVIEW

File: extensions/wikia/GlobalWatchlist/GlobalWatchlistBot.class.php [hunk 1]
```diff
@@ -9,7 +9,8 @@ class GlobalWatchlistBot {
 	const REPLY_ADDRESS = 'noreply@wikia.com';
 
 	public function __construct() {
-		\F::app()->wg->ExtensionMessagesFiles['GlobalWatchlist'] = dirname( __FILE__ ) . '/GlobalWatchlist.i18n.php';
+		$messageFiles = \F::app()->wg->ExtensionMessagesFiles;
+		$messageFiles['GlobalWatchlist'] = dirname( __FILE__ ) . '/GlobalWatchlist.i18n.php';
```
Comment (garthwebb) [inline extensions/wikia/GlobalWatchlist/GlobalWatchlistBot.class.php]: The line as is was throwing this error in dev:
```php
Indirect modification of overloaded property WikiaGlobalRegistry::$ExtensionMessagesFiles has no effect in /usr/wikia/source/app/extensions/wikia/GlobalWatchlist/GlobalWatchlistBot.class.php on line 12
```
Even though its just a notice, it was somehow randomly killing requests

File: extensions/wikia/ArticleComments/classes/ArticleComment.class.php [hunk 2]
```diff
@@ -13,18 +13,22 @@ class ArticleComment {
 	const CACHE_VERSION = 2;
```
Comment (garthwebb) [inline extensions/wikia/ArticleComments/classes/ArticleComment.class.php]: There are a lot of spacing fixes introduced in this class via the stylize.php script.  I also changed array() => [].  Sorry for the length ...

File: extensions/wikia/ArticleComments/classes/ArticleComment.class.php [hunk 3]
```diff
@@ -94,6 +98,52 @@ static public function newFromArticle( Article $article ) {
...
+		if ( empty( $param['useSlave'] ) ) {
+			$dbh = wfGetDB( DB_MASTER );
+			$flags = Title::GAID_FOR_UPDATE;
+		} else {
+			$dbh = wfGetDB( DB_SLAVE );
+			$flags = 0;
+		}
+
+		$titleText = $title->getDBkey();
+		$prefix =  $titleText . '/' . ARTICLECOMMENT_PREFIX;
```
Comment (rafalkalinski) [inline extensions/wikia/ArticleComments/classes/ArticleComment.class.php]: this is probably a good place for implode finction

Comment (garthwebb) [inline extensions/wikia/ArticleComments/classes/ArticleComment.class.php]: For joining two, possibly three things together I think simple concatenation is more readable than an implode.

File: extensions/wikia/Email/Email.i18n.php [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 $messages['en'] = array(
 	'emailext-desc' => 'An extension to handle delivering email',
+	'emailext-anonymous-editor' => 'A Wikia fan',
```
Comment (garthwebb) [inline extensions/wikia/Email/Email.i18n.php]: This message was moved from WatchedPage.i18n.php since it could be applied to multiple email types.

File: extensions/wikia/Email/EmailController.class.php [hunk 5]
```diff
@@ -28,6 +31,12 @@
 	/**	@var string The language to send the email in */
 	protected $targetLang;
```
Comment (garthwebb) [inline extensions/wikia/Email/EmailController.class.php]: The following two members were pulled down from the WatchedPageController since they are common to all emails sent.

File: extensions/wikia/Email/EmailController.class.php [hunk 6]
```diff
@@ -145,13 +173,24 @@ public function main() {
 	/**
 	 * Create an error response for any exception thrown while creating this email
 	 *
-	 * @param ControllerException $e
+	 * @param \Exception $e
 	 *
 	 */
-	protected function setErrorResponse( ControllerException $e ) {
+	protected function setErrorResponse( \Exception $e ) {
```
Comment (garthwebb) [inline extensions/wikia/Email/EmailController.class.php]: So it turns out that Nirvana change an exception thrown (say one based on our ```Email\ControllerException```) and re-throw it as its own ```WikiaDispatchedException```.  Given that we use the Nirvana request routing a few times during email creation, we need to catch all errors.

File: extensions/wikia/Email/templates/avatarLayout.mustache [hunk 7]
```diff
@@ -1,6 +1,6 @@
 <div class="greeting">
 	<p>{{salutation}}</p>
-	<p>{{{articleEditedText}}}</p>
+	<p>{{{summary}}}</p>
```
Comment (garthwebb) [inline extensions/wikia/Email/templates/avatarLayout.mustache]: Gave this and the variable below a more generic name, since the current names are somewhat specific to the WatchedPage email.

File: includes/EmailNotification.php [hunk 8]
```diff
@@ -458,30 +455,68 @@ private function emailExtensionEnabled() {
...
+		} elseif ( $this->isBlogComment() ) {
+			$controller = 'Email\Controller\BlogComment';
+		}
+
+		if ( !empty( $controller ) ) {
+			$params = [
 				'targetUser' => $user->getName(),
 				'title' => $this->title->getText(),
-				'nameSpace' => $this->title->getNamespace(),
+				'namespace' => $this->title->getNamespace(),
```
Comment (garthwebb) [inline includes/EmailNotification.php]: This was a merge conflict.  Kept the all lowercased version since ```namespace``` is one word.

File: includes/UserMailer.php [hunk 9]
```diff
@@ -146,21 +146,32 @@ static function makeMsgId() {
 	 * array of parameters. It requires PEAR:Mail to do that.
 	 * Otherwise it just uses the standard PHP 'mail' function.
 	 *
```
Comment (garthwebb) [inline includes/UserMailer.php]: Several fixes to doc strings ...

File: includes/UserMailer.php [hunk 10]
```diff
@@ -146,21 +146,32 @@ static function makeMsgId() {
...
+		MailAddress $to,
+		MailAddress $from,
+		$subject,
+		$body,
+		MailAddress $replyTo = null,
+		$contentType = null,
+		$category = 'UserMailer',
+		$priority = 0,
+		$attachments = []
+	) {
```
Comment (garthwebb) [inline includes/UserMailer.php]: Simply split this to several lines since its very long and hard to follow.

File: includes/UserMailer.php [hunk 11]
```diff
@@ -209,14 +220,14 @@ public static function send( $to, $from, $subject, $body, $replyto = null, $cont
 		$headers['From'] = $from->toString();
 		$headers['Return-Path'] = $from->address;
 
-		if ( $replyto && $replyto instanceof MailAddress ) {
-			$headers['Reply-To'] = $replyto->toString();
+		if ( $replyTo && $replyTo instanceof MailAddress ) {
+			$headers['Reply-To'] = $replyTo->toString();
```
Comment (garthwebb) [inline includes/UserMailer.php]: Applying casing rules.

File: includes/UserMailer.php [hunk 12]
```diff
@@ -209,14 +220,14 @@ public static function send( $to, $from, $subject, $body, $replyto = null, $cont
...
 		}
 
 		$headers['Date'] = date( 'r' );
 		$headers['MIME-Version'] = '1.0';
 
-		if ( empty( $attachements ) ) {
+		if ( empty( $attachments ) ) {
```
Comment (garthwebb) [inline includes/UserMailer.php]: Misspelled var.

File: extensions/wikia/GlobalWatchlist/GlobalWatchlistBot.class.php (hunk 1 above)
Comment (jsutterfield) [inline extensions/wikia/GlobalWatchlist/GlobalWatchlistBot.class.php]: Oh, hmm. Yea, when the bot is instantiated it's including the message files. Incorrectly at that. Looks like that was my change. Sorry about that. Thanks for fixing it.

But also, no, it doesn't really do anything. 1.) Because we're already including the file [here](https://github.com/Wikia/app/blob/dev/extensions/wikia/GlobalWatchlist/GlobalWatchlist.php#L21) during startup. And 2.) because most of these messages are in the MediaWiki namespace on messaging wiki which trumps this file. I'm going to remove that include from the set-up file since these messages are only need when the bot is run.

File: includes/context/RequestContext.php [hunk 13]
```diff
@@ -327,7 +327,7 @@ public function setSkin( Skin $s ) {
 	/**
 	 * Get the Skin object
 	 *
-	 * @return Skin
+	 * @return Skin|Linker
```
Comment (garthwebb) [inline includes/context/RequestContext.php]: The object returned is always of type ```Skin``` but this code contains a ```__call``` method that sends unmatched method calls to ```Linker``` methods.  Adding this hinting allows PHPStorm to find method definitions properly.  This kind of ```__call``` usage is essentially multiple inheritance.

File: includes/EmailNotification.php [hunk 14]
```diff
@@ -458,30 +455,68 @@ private function emailExtensionEnabled() {
...
 	private function sendUsingEmailExtension( \User $user ) {
-		F::app()->sendRequest( 'Email\Controller\WatchedPage', 'handle',
-			[
+
+		if ( $this->isArticlePageEdit() ) {
+			$controller = 'Email\Controller\WatchedPage';
+		} elseif ( $this->isArticleComment() ) {
+			$controller = 'Email\Controller\ArticleComment';
+		} elseif ( $this->isBlogComment() ) {
+			$controller = 'Email\Controller\BlogComment';
```
Comment (jsutterfield) [inline includes/EmailNotification.php]: So this doesn't actually cover all cases. Adding a category also goes through this method (`$this->action === "categoryadd"`), as does the email for a create page (`empty( $this->action ) && $this->isNewPage();`) so those emails will get skipped by this current logic. Those are the only other 2 that I know of, but given the wilyness of MediaWiki, there could be others.

We should make sure that the category add and created page emails still flows through the `sendUsingUserMailer()` method, and also maybe log the $action when the controller is empty on L471 so we can see if we're missing anything else?

File: extensions/wikia/Follow/FollowHelper.class.php [hunk 15]
```diff
@@ -220,7 +223,7 @@ static public function watchBlogListing(&$article, &$user, $text, $summary, $min
 				$title = Title::makeTitle( NS_BLOG_LISTING, $title );
 				$related[] = ucfirst($title->getDBKey());
 			}
-			self::emailNotification($article->getTitle(), $related, NS_BLOG_LISTING, $user, "blogpost", wfMsg('follow-bloglisting-summary'));
+			self::emailNotification($article->getTitle(), $related, NS_BLOG_LISTING, $user, self::LOG_ACTION_BLOG_POST, wfMsg('follow-bloglisting-summary'));
```
Comment (jsutterfield) [inline extensions/wikia/Follow/FollowHelper.class.php]: Since we're touching this line, how about updating wfMsg?

File: extensions/wikia/Email/i18n/Comment.i18n.php [hunk 16]
```diff
@@ -0,0 +1,18 @@
+<?php
+$messages = [];
+
+$messages['en'] = [
+	'emailext-articlecomment-summary' => 'People are talking about $1 on {{SITENAME}}!',
+	'emailext-blogcomment-summary' => 'People are talking about the $1 post on {{SITENAME}}!',
+	'emailext-comment-salutation' => 'Hi $1,',
+	'emailext-comment-link-label' => 'See this comment',
+	'emailext-comment-view-all' => "[$1 View all comments]",
```
Comment (jsutterfield) [inline extensions/wikia/Email/i18n/Comment.i18n.php]: Do we have a ticket to get these translated?

File: extensions/wikia/Email/EmailController.class.php [hunk 17]
```diff
@@ -344,6 +376,35 @@ private function getUnsubscribeLink() {
 		return $unsubscribeTitle->getFullURL( $params );
 	}
 
+
+	/**
+	 * @return String
+	 */
+	protected function getCurrentProfilePage() {
+		if ( $this->currentUser->isAnon() ) {
```
Comment (jsutterfield) [inline extensions/wikia/Email/EmailController.class.php]: Don't we want to check if the currentUser is not anon?

File: extensions/wikia/Email/EmailController.class.php [hunk 18]
```diff
@@ -344,6 +376,35 @@ private function getUnsubscribeLink() {
...
+			return $this->currentUser->getUserPage()->getFullURL();
+		}
+		return "";
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getCurrentUserName() {
+		if ( $this->currentUser->isAnon() )	 {
```
Comment (jsutterfield) [inline extensions/wikia/Email/EmailController.class.php]: Same here, shouldn't this be `!$this->currentUser->isAnon()`?

File: extensions/wikia/Email/Email.i18n.php (hunk 4 above)
Comment (mklucsarits) [inline extensions/wikia/Email/Email.i18n.php]: Can the capitalization issue in https://wikia-inc.atlassian.net/browse/SOC-641 be handled as part of your changes here?

Comment (garthwebb) [inline extensions/wikia/Email/Email.i18n.php]: @mklucsarits good point.  I'm on the edge of whether to do it though since it seems like she may want two different behaviors out this string which I'm inclined to handle separately.  I'll see what she says.

File: extensions/wikia/Email/EmailController.class.php (hunk 17 above)
Comment (garthwebb) [inline extensions/wikia/Email/EmailController.class.php]: Yup.  I don't really remember why I changed this from ```isLoggedIn```

File: extensions/wikia/Email/i18n/Comment.i18n.php (hunk 16 above)
Comment (garthwebb) [inline extensions/wikia/Email/i18n/Comment.i18n.php]: No, I need to do that.

Comment (garthwebb) [inline extensions/wikia/Email/i18n/Comment.i18n.php]: And now, witness the glory that is LOC-28: https://wikia-inc.atlassian.net/browse/LOC-28

File: extensions/wikia/Follow/FollowHelper.class.php (hunk 15 above)
Comment (garthwebb) [inline extensions/wikia/Follow/FollowHelper.class.php]: Sure, and while I'm at it I'll UPDATE SPACING FOR THE WHOLE FILE!  And add some comments.  While I'm at it.

File: includes/EmailNotification.php (hunk 14 above)
Comment (garthwebb) [inline includes/EmailNotification.php]: Yah, that was my intent but I think I lost it when moving things around.

File: includes/EmailNotification.php [hunk 19]
```diff
@@ -496,44 +541,45 @@ private function sendUsingUserMailer( \User $user ) {
...
 	private function sendPersonalised( $watchingUser ) {
-		// From the PHP manual:
-		//     Note:  The to parameter cannot be an address in the form of "Something <someone@example.com>".
-		//     The mail command will not parse this properly while talking with the MTA.
 		$to = new MailAddress( $watchingUser );
-
-		// $PAGEEDITDATE is the time and date of the page change
-		// expressed in terms of individual local time of the notification
-		// recipient, i.e. watching user
-		// TODO The logic below is duplicated twice, break this off into a separate method.
```
Comment (garthwebb) [inline includes/EmailNotification.php]: Actually, it was only duplicated once ...

File: extensions/wikia/Follow/FollowHelper.class.php (hunk 15 above)
Comment (jsutterfield) [inline extensions/wikia/Follow/FollowHelper.class.php]: LOL

**Comment by jsutterfield** [comment]
LGTM

File: extensions/wikia/Email/Email.i18n.php (hunk 4 above)
Comment (garthwebb) [inline extensions/wikia/Email/Email.i18n.php]: @mklucsarits ok, I'm going to let this one be for this ticket and not confuse things.

File: extensions/wikia/Email/i18n/Comment.i18n.php (hunk 16 above)
Comment (garthwebb) [inline extensions/wikia/Email/i18n/Comment.i18n.php]: Ok, LOC-28 got ... lost in translation.  Here's the new ticket, LOC-32: https://wikia-inc.atlassian.net/browse/LOC-32

File: includes/EmailNotification.php (hunk 14 above)
Comment (garthwebb) [inline includes/EmailNotification.php]: Ok, I think this should be addressed by the changes made in the calling function.

File: extensions/wikia/Follow/FollowHelper.class.php (hunk 15 above)
Comment (garthwebb) [inline extensions/wikia/Follow/FollowHelper.class.php]: Hmm, somehow with all this I didn't actually update wfMsg...
